scripts/staking/snapshot_staking.ts

```diff
@@ -0,0 +1,201 @@
+// scripts/staking/snapshot-staking.ts
+import { ethers } from "ethers";
+import { createObjectCsvWriter } from "csv-writer";
+import fs from "fs";
+import path from "path";
+import { TBOS_SNAPSHOT_STAKING_CONFIG } from "./config/tbos_snapshot_staking.config";
+
+interface StakingSnapshot {
+  address: string;
+  amountStaked: string;
+  votingPower: string;
+  isVesting: boolean;
+}
+
+const STAKING_ABI = [
+  "function getCurrentVotes(address account) external view returns (uint96)",
+  "function getPriorVotes(address account, uint256 blockNumber, uint256 date) public view returns (uint96)",
+  "function isVestingContract(address stakerAddress) external view returns (bool)",
+  "function _currentBalance(address account, uint256 lockDate) internal view returns (uint96)",
+  "function getPriorStakeByDateForDelegatee(address account, uint256 date, uint256 blockNumber) external view returns (uint96)",
+  "function balanceOf(address account) external view returns (uint96)",
+  "event TokensStaked(address indexed staker, uint256 amount, uint256 lockedUntil, uint256 totalStaked)",
+];
+
+async function getStakingSnapshot(params: {
+  rpcUrl: string;
+  stakingAddress: string;
+  stakerAddresses: string[];
+  snapshotTimestamp?: number;
+  voluntaryOnly?: boolean;
+}) {
+  const {
+    rpcUrl,
+    stakingAddress,
+    stakerAddresses,
+    snapshotTimestamp,
+    voluntaryOnly = false,
+  } = params;
+  console.log("Starting staking snapshot process...");
+
+  const provider = new ethers.providers.JsonRpcProvider(rpcUrl);
+  const stakingContract = new ethers.Contract(
+    stakingAddress,
+    STAKING_ABI,
+    provider,
+  );
+
+  // Get current block if no timestamp provided
+  const currentBlock = await provider.getBlock("latest");
+  const targetTimestamp = snapshotTimestamp || currentBlock.timestamp;
+
+  // Calculate target block number based on timestamp
+  const targetBlock = await calculateBlockNumber(
+    provider,
+    TBOS_SNAPSHOT_STAKING_CONFIG.averageBlockTime,
+    targetTimestamp,
+  );
+  console.log(
+    `Target timestamp: ${targetTimestamp} (${new Date(targetTimestamp * 1000).toISOString()})`,
+  );
+  console.log(`Target block: ${targetBlock}`);
+
+  const results: StakingSnapshot[] = [];
+  let processedCount = 0;
+
+  for (const staker of stakerAddresses) {
+    try {
+      // Check if it's a vesting contract
+      const isVesting = await stakingContract.isVestingContract(staker);
+
+      // Skip if voluntaryOnly is true and this is a vesting contract
+      if (voluntaryOnly && isVesting) {
```

Comment:
this is a mixed up logic - should be divided.  
voluntaryOnly means that only voluntary staking is taken into account and is in effect == total VP sub delegated VP (delegated VP includes VP delegated by vesting smart contracts, but also can include delegation from other stakers),  
and `isVesting` requires check if the address is not vesting smart contract. 

```suggestion
      if (isVesting) {
```

Reply:
Updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/555/commits/2fa2b59c9cd238a5103b649ecc26a334bd0b208d)